ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,287 @@
+}
+
+func parseInstanceID(instanceID string) (int, error) {
+	parts := instanceIDRegex.FindStringSubmatch(instanceID)
+	if len(parts) != 3 {
+		return 0, errorBadInstanceIDFormat(instanceID)
+	}
+	return strconv.Atoi(parts[2])
+}
+
+// findZoneID gets a zone name and a slice of sorted zones,
+// and return the index of the zone in the slice.
+func findZoneID(zone string, sortedZones []string) (int, error) {
+	index := slices.Index(sortedZones, zone)
+	if index < 0 {
+		return 0, errorZoneNotValid(zone)
+	}
+	return index, nil
+}
+
+// generateFirstInstanceTokens calculates a set of tokens that should be assigned to the first instance (with id 0)
+// of the zone of the underlying instance.
+func (t *SpreadMinimizingTokenGenerator) generateFirstInstanceTokens() Tokens {
+	tokenDistance := (totalTokensCount / optimalTokensPerInstance / maxZonesCount) * maxZonesCount
+	tokens := make(Tokens, 0, optimalTokensPerInstance)
+	for i := 0; i < optimalTokensPerInstance; i++ {
+		token := uint32(i*tokenDistance) + uint32(t.zoneID)
+		tokens = append(tokens, token)
+	}
+	return tokens
+}
+
+// calculateNewToken determines where in the range represented by the given ringToken should a new token be placed
+// in order to satisfy the constraint represented by the optimalTokenOwnership. This method assumes that:
+// - ringToken.token % zonesCount == ringToken.prevToken % zonesCount
+// - optimalTokenOwnership % zonesCount == 0,
+// where zonesCount is the number of zones in the ring. The caller of this function must ensure that these assumptions hold.
+func (t *SpreadMinimizingTokenGenerator) calculateNewToken(token ringToken, optimalTokenOwnership uint32) (uint32, error) {
+	if optimalTokenOwnership < maxZonesCount || optimalTokenOwnership%maxZonesCount != 0 {
+		return 0, errorMultipleOfZonesCount(optimalTokenOwnership, token)
+	}
+	if token.prevToken%maxZonesCount != token.token%maxZonesCount {
+		return 0, errorLowerAndUpperBoundModulo(optimalTokenOwnership, token)
+	}
+	ownership := tokenDistance(token.prevToken, token.token)
+	if ownership <= int(optimalTokenOwnership) {
+		return 0, errorDistanceBetweenTokensNotBigEnough(int(optimalTokenOwnership), ownership, token)
+	}
+	maxTokenValue := (math.MaxUint32/uint32(maxZonesCount) - 1) * maxZonesCount
+	offset := maxTokenValue - token.prevToken
+	if offset < optimalTokenOwnership {
+		newToken := optimalTokenOwnership - offset
+		return newToken, nil
+	}
+	return token.prevToken + optimalTokenOwnership, nil
+}
+
+// optimalTokenOwnership calculates the optimal ownership of the remaining currTokensCount tokens of an instance
+// having the given current instances ownership currInstanceOwnership and the given optimal instance ownership
+// optimalInstanceOwnership. The resulting token ownership must be a multiple of the number of zones.
+func (t *SpreadMinimizingTokenGenerator) optimalTokenOwnership(optimalInstanceOwnership, currInstanceOwnership float64, currTokensCount uint32) uint32 {
```

Comment:
[nit] Have you considered rename "currTokensCount" to "remainingTokensCount"? I think may help to read it.